### ISSUE TITLE
Prefer composition over inheritance

### ISSUE THREAD

**Issue by kopacki** [ryanmcdermott/clean-code-javascript#40]
As [the rule](https://medium.com/javascript-scene/3-different-kinds-of-prototypal-inheritance-es6-edition-32d777fa16c9#79a1) from title tells us, we should favor composition when we code.
So maybe it's worth to replace all examples using `class` with proper ones based on composition? 

Let's take the ["Avoid conditionals"](https://github.com/ryanmcdermott/clean-code-javascript#avoid-conditionals) section for instance. It shows the example of inheritance while the goal was to encourage the developer to use composition. Below you can see how the definition of different airplanes could look like (using composition):
```javascript
const Airplane = {
  getName() {
    return `${this.name}`
  }
}

const AirbusA380 = { 
  name: 'Airbus A380',
  getName() {
    return  `${Airplane.getName.call(this)} of ${this.airline}`
  }
}

const Cessna = { 
  name: 'Cessna'
}


// define factory functions (airbusA380, cessna)
const airbusA380 = (state) => Object.assign(
  {},
  Airplane,
  AirbusA380,
  state
)

const cessna = (state) => Object.assign(
  {},
  Airplane,
  Cessna,
  state
)

// create instances of airplanes
const emiratesAirbus = airbusA380({
  airline: 'Emirates'
})

const privateCessnaJet = cessna()

console.log(emiratesAirbus.getName()) // prints "Airbus A380 of Emirates"
console.log(privateCessnaJet.getName()) // prints "Cessna"
```

What do you think about replacing all `class` examples with `composition-based` ones?

**Comment by cookiengineer** [comment]
The issue #45 seems to be also relevant, as the example with `withdraw(amount)` seems to imply Composition Pattern.

**Comment by ericelliott** [comment]
:+1:

**Comment by Tarabass** [comment]
@ericelliott When I saw the example I almost instantly thought about you. I wanted to open an issue as well about this, but good not find your post on medium about why not extend classes. Maybe you can post that link here so we can learn?

**Comment by jokeyrhyme** [comment]
@gaearon has some great advice about using `class`:

https://medium.com/@dan_abramov/how-to-use-classes-and-sleep-at-night-9af8de78ccb4#.93lw502p8

It mostly boils down to limiting use of `extend` and `super`, and still preferring composition over inheritance.

If I was working in an existing codebase with a traditional classy architecture, I'd certainly prefer to use `class` instead of the weird hoop-jumping we had to do in ES5 and older.

**Comment by dschinkel** [comment]
Why use class all the time.  Why not code composition with Object.create() instead which doesn't couple you to everything like inheritance does?

**Comment by cookiengineer** [comment]
@dschinkel Instead of `Object.create()` clutter, you can also use a simple `Object.assign()` in combination with `mycomposite.prototype = Object.assign({}, other)` :)

**Comment by aaronmcadam** [comment]
I think changing this part of the guide is tricky because the traditional *[replace conditional with polymorphism](https://sourcemaking.com/refactoring/replace-conditional-with-polymorphism)* refactoring uses *[template method](https://sourcemaking.com/design_patterns/template_method)* which uses inheritance. Maybe using composition here would confuse more than it would teach.

**Comment by ericelliott** [comment]
There are several options for inheritance in JavaScript, described in detail in ["What's the Difference Between Class and Prototypal Inheritance?](https://medium.com/javascript-scene/master-the-javascript-interview-what-s-the-difference-between-class-prototypal-inheritance-e4cd0a7562e9#.exxs3ldyo)

**Comment by dschinkel** [comment]
cool thx for the share @ericelliott

**Comment by dschinkel** [comment]
btw @ericelliott just so you know, a friend of mine isn't fond of prototypal inheritance in terms of the new() constructor (he uses Object.create() instead).  I also have read up on "You don't know JS" and he has some good reasons to use Object.create() (OLOO style code).

https://github.com/getify/You-Dont-Know-JS/blob/master/this%20&%20object%20prototypes/README.md#you-dont-know-js-this--object-prototypes

https://github.com/getify/You-Dont-Know-JS/blob/master/this%20%26%20object%20prototypes/ch6.md

Now coming from the past, being in a language that's typed such as C#, I hated abstract classes.  It was code smell when you go to a place where they try to make everything based on inheritance (Abstract base classes) over composition (interfaces / decoupled).

Looking at JS, I kinda would look at the use of prototypal inheritance as code smell, that's just personal rather than just compose and use objects (decoupled).

Thoughts?

I'm not here to start a debate, or to change the subject, but it's been something on my mind lately.

**Comment by dschinkel** [comment]
@Tarabass 
> ut does it really matter what we use? 

yes to a degree.  In any language, there are debates but there are good practices that have clearly been agreed to upon most the community.  Lets say I was writing a "clean-code-csharp".  Would I use abstract classes everywhere in examples, no way.  That's clearly code smell.  Not something I'd want to be promoting people to be adopting.  There are way too many .NET shops still thinking that using Abstract classes throughout their entire codebase is the way to go.  And you end up with a coupled mess.

**Comment by dschinkel** [comment]
I also think  having conversations around this repo about coding style is very approrpriate and a relly good place to be doing so.  After-all, there's a Clean Code forum in Google Forums where people talk about syntax a lot as it relates to Clean Code principals.  So I feel people giving inputs in here in comments is just fine.

**Comment by ericelliott** [comment]
@dschinkel `Object.create()` is just one of a few ways to use prototypal inheritance in JavaScript using prototype **delegation**.

`class extends` is another way to accomplish a similar thing.

Constructor functions and `new` is another way to accomplish the same thing.

All of them create delegation links using OLOO (Objects Linking to Other Objects). Read the links I posted above for more details as well as ["Common Misconceptions About Inheritance in JavaScript"](https://medium.com/javascript-scene/common-misconceptions-about-inheritance-in-javascript-d5d9bab29b0a) for much more on the topic.

**Comment by dschinkel** [comment]
Great post, thx!

**Comment by kopacki** [comment]
@Tarabass @aaronmcadam I think each programming language has its own clean code way - it mainly depends on features it has. As @ericelliott mentioned JavaScript inheritance finally comes to OLOO (Objects Linking to Other Objects). The `class` keyword is a syntax sugar and it suggest you can use a classical inheritance in your code, but in fact you can't.

Of course you can still use it as it does its job, although it's not the best option - and since we try here to define proper clean code guide for JavaScript we should promote the best options first.

I believe the inheritance/composition topic is comprehensively covered here:
- ["Common Misconceptions About Inheritance in JavaScript"](https://medium.com/javascript-scene/common-misconceptions-about-inheritance-in-javascript-d5d9bab29b0a)
- ["3 Different Kinds of Prototypal Inheritance: ES6+ Edition"](https://medium.com/javascript-scene/3-different-kinds-of-prototypal-inheritance-es6-edition-32d777fa16c9)
- ["What's the Difference Between Class and Prototypal Inheritance?"](https://medium.com/javascript-scene/master-the-javascript-interview-what-s-the-difference-between-class-prototypal-inheritance-e4cd0a7562e9)
- [You Don't Know JS: Simpler design](https://github.com/getify/You-Dont-Know-JS/blob/master/this%20%26%20object%20prototypes/ch6.md#simpler-design)

**Comment by dschinkel** [comment]
thx for the share @kopacki

**Comment by Jeff-Mott-OR** [comment]
@kopacki, @Tarabass, @dschinkel

I hate to be the wet blanket, but unfortunately Eric Elliott got his information wrong.

### Invented definitions

In Elliott's blogs, he literally invented his own personal definition of object composition, and his invented definition is different than how the GoF used the term when they coined the phrase "favor composition". Elliott even cited Wikipedia's article on object composition just so he could rant about how he thinks Wikipedia got it wrong. To say we should favor composition is only meaningful if we all agree what that term means, and Elliott's personally invented meaning is in conflict with the GoF and Wikipedia and MDN and damn near every other reference.

### It's actually inheritance

What Elliott calls composition is in fact still inheritance. This became especially obvious when, bizarrely, Elliott claimed that `class A extends B` is composition, even in an unambiguously classical language such as C++. I shit you not.

### Multiple inheritance

The killer feature of what Elliott mistakenly calls composition is in fact just multiple inheritance. Python and C++, for example, are a couple languages that natively support multiple inheritance, so the airplane code above could be written in Python like this:
    
    class Airplane:
      def getName(self):
        return self.name
    
    class AirbusA380:
      name = "Airbus A380"
      def getName(self):
        return Airplane.getName(self) + " of " + self.airline
    
    class Cessna:
      name = "Cessna"
    
    # this is multiple inheritance, not composition
    class airbusA380(AirbusA380, Airplane):
      def __init__(self, state = dict()):
        self.__dict__.update(state)
    
    class cessnaJet(Cessna, Airplane):
      def __init__(self, state = dict()):
        self.__dict__.update(state)
    
    # create instances of airplanes
    emiratesAirbus = airbusA380(dict({
      "airline": "Emirates"
    }))
    
    privateCessnaJet = cessnaJet()
    
    print(emiratesAirbus.getName()) # prints "Airbus A380 of Emirates"
    print(privateCessnaJet.getName()) # prints "Cessna"

### Not immune

Elliott frequently claims that what he mistakenly calls composition is immune to all the problems that plague inheritance, but in fact it isn't immune to any of them -- which shouldn't be surprising now that we know it's actually still inheritance. The airline code above, for example, is in fact still susceptible to the fragile base problem. Let's say the Airplane object is supplied by a library, and the rest is your own code that uses that Airplane object. Then one day the maintainer puts out a new release, and now the Airplane object looks like this:

    const Airplane = {
      getName() {
        this.airline();
        return `${this.name}`
      },
      airline() {}
    }

The maintainer of the Airplane object is good about semantic versioning, so he updates the **MINOR** version number, because, after all, from his perspective, this change is fully backward compatible. And yet, even though it's fully backward compatible, it will nonetheless break your code. This is the fragile base problem, an issue that we normally associate with inheritance, and yet here it is.

### tl;dr

If you learned about composition or inheritance from Eric Elliott, then everything you think you know is wrong.

**Comment by dschinkel** [comment]
I'm waiting for the gloves to be put on :).

**Comment by ericelliott** [comment]
@Jeff-Mott-OR I addressed this nonsense in ["The Open Minded Explorer's Guide to Object Composition"](“The Open Minded Explorer’s Guide to Object Composition"](https://medium.com/javascript-scene/the-open-minded-explorer-s-guide-to-object-composition-88fe68961bed).

Basically, somebody got it in their heads that the GoF composite pattern is the be-all and end-all of object composition and no other form of composite object building can qualify as "object composition" and started spreading a bunch of misconceptions.

Complete and utter nonsense.

Yes, object composition is a form of inheritance. That's why object concatenation is often called **concatenative inheritance**. Yes, concatenative inheritance is a form of multiple inheritance, but concatenative inheritance does not necessarily create class hierarchies, unlike multiple inheritance in C++, for example, where it's possible to run into the diamond problem when there are property name collisions.

Yes, it's possible to build class hierarchies using prototypal inheritance techniques. That's what's happening under the hood when you mimic class inheritance in JavaScript. (I.e., `class foo extends bar`).

Immunity from the fragile base class problem only comes when you avoid hierarchies. The distinctions are made clear in ["Master the JavaScript Interview: What’s the Difference Between Class & Prototypal Inheritance?"](https://medium.com/javascript-scene/master-the-javascript-interview-what-s-the-difference-between-class-prototypal-inheritance-e4cd0a7562e9).

Some people are more interested in winning arguments on the interwebs than trying to understand each other, and so tied to their narrow understanding of a word that they can't see broader applications.

Some people think if you use a word that violates their narrow definition of it, that means literally everything you say about it is wrong.

That is absolute nonsense, of course.

The whole paragraph that ends with "I shit you not" above is such an intentional distortion that I laughed out loud when I got to that part.

Pissing contests aside, composition over class inheritance boils down to rigid hierarchy vs flexibility.

Disagree with the words I use to describe things all you like. Just pick the techniques that lead to the most flexibility and you'll be all right. I don't care if you chose to call it multiple inheritance. Just don't try to pretend that building out a deep hierarchy with `extends` is a good practice. ;)

"Favor object composition over class inheritance" is a means to an end: build more flexible code.

**Comment by Jeff-Mott-OR** [comment]
The good news is that Elliott acknowledged for everyone here that what he calls object composition is actually inheritance.

It turns out that what he calls object composition is inheritance plus a couple rules that Elliott considers a best practice, and what he calls class inheritance is inheritance that does not follow those couple rules.

### Inherit only one level deep

The first of Elliott's personal rules is that we must inherit only one level deep.

    class A extends B {}

This, according to Eric Elliott, is object composition and not class inheritance. But if we inherit two levels deep...

    class B extends C {}
    class A extends B {}

Only now -- again, according to Eric Elliott -- does this become class inheritance.

This rule is at least *somewhat* close to a real, widely accepted best practice -- that we should prefer shallow hierarchies. But there's no hard limit on one level, and even shallow hierarchies are still hierarchies. Contrary to Elliott's claims, both of the above code examples are inheritance and neither are object composition.

### No overrides

The second of Elliott's rules is that the derived type must not override anything from the base type.

    class B {
        fn() {}
    }

    class A extends B {}

This, according to Eric Elliott, is object composition and not class inheritance. But if we override something...

    class B {
        fn() {}
    }

    class A extends B {
        // Override
        fn() {}
    }

Only now -- again, according to Eric Elliott -- does this become class inheritance.

But this time there is no industry-accepted best practice that matches Elliott's personal rule. He thinks following this rule will save you from the fragile base problem, but actually it won't. Third parties can update their libraries in ways that are fully backward compatible and still break your code.

### Definitions matter

When the GoF said to favor composition, of course they didn't mean you should code however you please just so long as you label it composition. Eric Elliott is advocating for a very different technique -- in fact the literal *opposite* technique -- than what the GoF were arguing for. Folks who learn from Eric Elliott are lead to believe they're favoring composition but actually they're abusing inheritance more than ever.

### tl;dr

Eric Elliott sold you blue water and called it medicine.

**Comment by dschinkel** [comment]
correction to my previous comment.  My friend prefers not to use Object.Create() and uses Object.assign() as he says it's very useful for composition and doesn't like chaining...there are other patterns that are more useful he says than using chaining is not needed at all.  He said he's good with composition chains which would be a whole other conversation.  He agrees with Eric Elliot and kopacki and has been doing JS just as long as anyone in here.

**Comment by jokeyrhyme** [comment]
> Eric Elliott sold you blue water and called it medicine.

@Jeff-Mott-OR let's try to keep things civil and respectful. You can discuss the merits or pitfalls of an idea without maligning the character of someone else.

It's obvious that you are passionate about this discussion, let's just try to be passionate _and_ constructive.

**Comment by jokeyrhyme** [comment]
> This, according to Eric Elliott, is object composition and not class inheritance.

@Jeff-Mott-OR none of your code examples in https://github.com/ryanmcdermott/clean-code-javascript/issues/40#issuecomment-275515745 are what @ericelliott describes as "object composition". To me (at least), the second the `extends` keyword or the `prototype` property is involved, we are talking about inheritance, not composition.

**Comment by ryanmcdermott** [comment]
I echo what @jokeyrhyme says above, let's keep it civil! Very good to hear a thorough discussion about these topics from very experienced developers such as yourselves, we can come to a friendly resolution 😄

**Comment by Jeff-Mott-OR** [comment]
> none of your code examples in #40 (comment) are what @ericelliott describes as "object composition". 

Eric Elliott himself described them as such. Here is [a question](https://medium.com/@xwt23v/i-m-genuinely-trying-to-figure-out-where-you-draw-the-line-between-inheritance-and-what-you-call-ab11d70f7aaa) asked to Elliott, and [his reply](https://medium.com/@_ericelliott/again-all-semantics-b52e029c5df8) (plus an [imgur mirror](https://imgur.com/a/vA3T7) of his reply in case it gets deleted, which has happened before).

**Comment by jokeyrhyme** [comment]
Thanks for the links / screenshot, my favourite part is this:

> Forget which paintbrush you use to paint with. Look at the painting, instead.

Regardless of the terminology we use, ultimately we're discussing how best to share code between objects in a way that minimises copy-pasting / duplication. Ideally, we only want to fix a bug once, not across 9 different copies of the same function and mistakenly forgetting the 10th copy.

Besides maintenance issues, traditional inheritance features often exemplify the [gorilla-banana problem](https://www.johndcook.com/blog/2011/07/19/you-wanted-banana/), and techniques that we sometimes refer to as "composition" offer an alternative approach that grants you the banana without mandating an accompanying gorilla.

**Comment by Jeff-Mott-OR** [comment]
> Regardless of the terminology we use, ultimately we're discussing how best to share code between objects in a way that minimises copy-pasting / duplication.

Sure. Just be aware that the paintbrush Eric Elliott advocates to share code is **inheritance**. At no point did you learn to use object composition. Nor does Elliott's inheritance-in-disguise solve any of the problems he claims it does. You're still susceptible to the fragile base problem, and you can still get yourself into the banana-gorilla problem if you're not careful.

**Comment by jokeyrhyme** [comment]
In React, we use "higher-order components". Maybe another way to think of "composition" is "higher-order factories", or "higher-order constructors".

**Comment by ericelliott** [comment]
@Jeff-Mott-OR I am in this thread, and I can speak for myself. You've already tried to explain my position a couple times, and both times completely missed the major, salient difference between composition and **class** inheritance (remember, there are several different kinds of inheritance).

## So, for the record:

**I didn't invent any of these definitions. This stuff has been around in some form or other since before I was born.**

**Class** inheritance creates an **is-a relationships**, which creates class taxonomies -- essentially an inheritance tree. In JavaScript, this is largely accomplished with **prototype delegation** -- a form of prototypal inheritance.

**Object composition** creates **has-a** or **uses-a** relationships. Instead of B is A, B **has an** A or **uses an** A. In JavaScript, this is largely accomplished with concatenative inheritance (AKA mixins), a form of prototypal inheritance.

The central **is-a vs has-a distinction** is explained well by -- well hell, just **Google this for yourself** -- you'll find lots of examples of people saying similar things: "what's the difference between composition and inheritance?"

Note that lots of the answers are about Java, and nobody in the Java world listens to me, so there's **no way my influence has snuck in and distorted definitions over there**.

People are too lazy to Google, so I'll just quote a few:

> "They are absolutely different. Inheritance is an 'is-a' relationship. Composition is a 'has-a'."
[source: StackOverflow, top answer, first sentence, "Difference between Inheritance and Composition"](http://stackoverflow.com/questions/2399544/difference-between-inheritance-and-composition)

> "Make sure inheritance models the is-a relationship My main guiding philosophy is that inheritance should be used only when a subclass is-a superclass. In the example above, an Apple likely is-a Fruit, so I would be inclined to use inheritance. [...] Don't use inheritance just to get code reuse If all you really want is to reuse code and there is no is-a relationship in sight, use composition." - Java World, ["Inheritance versus composition: Which one should you choose?"](http://www.javaworld.com/article/2076814/core-java/inheritance-versus-composition--which-one-should-you-choose-.html)

> Composition - has-a relationship between objects.
> Inheritance - is-a relationship between classes. - [Quora top answer "What is the difference between composition and inheritance in Java?"](https://www.quora.com/What-is-the-difference-between-composition-and-inheritance-in-Java)

What does this mean in practice in JavaScript? For a really nice explanation using JavaScript watch ["Composition Over Inheritance"](https://www.youtube.com/watch?v=wfMtDGfHWpA).

## Composition vs Inheritance Boils Down to Is-A vs Has-A

Is-A vs Has-A/Uses-A is the primary salient difference between **class inheritance** and **object composition**. All this dribbling on about techniques like `class A extends B` vs `Object.create()` vs `Object.assign()` is **completely, utterly, blindingly beside the point.**


## Why Not Just Use the Composite Pattern Like the GoF was Talking About?

Wouldn't that be nice? Then we could all stop throwing mud at each other over strict stickler definitions and get some programming done.

It turns out, there's a good reason not to use the composite pattern in JavaScript: It's more complicated than necessary if all you want to do is achieve `foo` **has feature** `bar`. So why do they use it so much more in Java? Because **Java has static type classes**. Extending objects in Java requires a lot more jumping through hoops. **The GoF were Java programmers writing for an audience who mostly understood Java.** This is an important bit of historical context that *people who misunderstand composition* seem to be missing.

In Java, composition is frequently accomplished with the composite pattern or delegate objects because those are the most convenient ways to inherit using **feature-based rather than taxonomy based inheritance**.

Because JavaScript supports dynamic object extension, concatenation/mixins can accomplish the same goal (added flexibility, through **feature-based rather than taxonomy-based inheritance**) by extending the target object, instead. This leads to less jumping through hoops, and works for 99% of cases. Once in a while, you want to encapsulate the features that the new object will have and expose a different public API, in which case, the composite pattern / method delegation are fine in JavaScript, too.

Again, to restate my thesis from above, **I don't care** what **techniques** people use to accomplish **has-a**/**uses-a** relationships instead of **is-a** relationships. That's completely, absolutely, unequivocally **missing the point.**

**Comment by Jeff-Mott-OR** [comment]
> Object composition creates has-a or uses-a relationships. Instead of B is A, B has an A or uses an A.

That is correct.

> In JavaScript, this [Object composition] is largely accomplished with concatenative inheritance (AKA mixins), a form of prototypal inheritance.

That is ***not*** correct. 

Let's look at [Wikipedia for IS-A](https://en.wikipedia.org/wiki/Is-a).

> Type A **is a** subtype of type B when A’s speciﬁcation implies B’s speciﬁcation. That is, any object (or class) that satisﬁes A’s speciﬁcation also satisﬁes B’s speciﬁcation

When we do `a = Object.assign({}, b)` or `a = stampit.compose(b)`, then "a" can do everything "b" can do. "a" satisfies "b"'s' specification. An "a" IS-A "b".

There are also code samples on that Wikipedia page. For example:

    class A:
        def doSomethingALike(self):
            pass

    class B(A):
        def doSomethingBLike(self):
            pass

This is inheritance one level deep with no overrides. You've explicitly said in past discussions that you consider this to be object composition, yet here's Wikipedia explicitly saying this is an IS-A relationship.

Of course, you've already said in your blog how you think Wikipedia got object composition wrong, so I have no doubt you think Wikipedia got IS-A wrong as well. So let's also look at the very links you just supplied:

> You do composition by having an instance of another class C as a field of your class

> About composition By composition, I simply mean using instance variables that are references to other objects. For example: `class Apple { private Fruit fruit = new Fruit(); }`

**Comment by ericelliott** [comment]
It's clear you will never be satisfied by any version of composition other than the composite pattern. That's fine with me. You do you.

For any readers actually interested in learning and building software with flexible code, you can read about how object concatenation solves the fragile base class problem [here](https://medium.com/javascript-scene/master-the-javascript-interview-what-s-the-difference-between-class-prototypal-inheritance-e4cd0a7562e9).

Or watch the very good video I linked to in the previous message.

@Jeff-Mott-OR, whenever I say "composition", you have my permission to replace the word inside your head with whatever word helps you sleep better at night.

Let's let these nice people get on with their lives now.

**Comment by dschinkel** [comment]
Agree, while debates are healthy there is more than enough experience and content out there for people to make their own conclusions.  In addition to Eric's posts, I also read You Don't Know JS.  With Eric and that, that's more than enough for me to make conclusions on "composition".  Lets move on.

**Comment by Jeff-Mott-OR** [comment]
> I also read You Don't Know JS. With Eric and that, that's more than enough for me to make conclusions on "composition".

I didn't recall YDKJS mentioning composition very often, so I searched it and found this:

>
    var auth = new AuthController(
        // in addition to inheritance, we also need composition
        new LoginController()
    );
    auth.checkAuth();

> The other thing to mention is that we chose some composition to sprinkle in on top of the inheritance. AuthController needs to know about LoginController, so we instantiate it (new LoginController()) and keep a class member property called this.login to reference it, so that AuthController can invoke behavior on LoginController.

Turns out YDKJS's description and implementation of composition matches what I've been describing this entire time. And matches what the GoF described. And matches what Wikipedia described. And matches what the SO links above described.

Just about everyone in the industry has the same understanding of what object composition is... except Eric Elliott.

**Comment by ericelliott** [comment]
Fine. Fine. Call concatenation "duct-tape inheritance" for all I care. Re-read [this article](https://medium.com/javascript-scene/master-the-javascript-interview-what-s-the-difference-between-class-prototypal-inheritance-e4cd0a7562e9) and replace every instance of "composition" with "duct-tape inheritance".

It's all the same to me.

**Comment by dschinkel** [comment]
I want my mommy.

**Comment by dschinkel** [comment]
I vote to close this ticket @ryanmcdermott

**Comment by ryanmcdermott** [comment]
Closing this Issue for now, thanks for everyone's contributions :smile:
